test/unit/pudl_environment_test.py

```diff
@@ -18,173 +15,14 @@ def setup():
-        del os.environ["PUDL_OUTPUT"]
-    if os.getenv("PUDL_INPUT"):
-        del os.environ["PUDL_INPUT"]
-
-    workspace = "/test/whatever"
-    default_settings = {
-        "pudl_in": workspace,
-        "pudl_out": workspace,
-    }
-
-    settings_yaml = StringIO(yaml.dump(default_settings))
-
-    settings = get_defaults(
-        yaml_file=settings_yaml, output_dir=pudl_output_tmpdir / "output"
-    )
-
-    expected_values = {
-        "pudl_in": f"{workspace}",
-        "pudl_out": f"{pudl_output_tmpdir}/output",
-        "data_dir": f"{workspace}/data",
-    }
-
-    for key, value in expected_values.items():
-        assert (key, settings[key]) == (key, value)
-
-    assert os.getenv("PUDL_OUTPUT") == f"{pudl_output_tmpdir}/output"
-    assert os.getenv("PUDL_INPUT") == f"{default_settings['pudl_in']}/data"
-
-
-@pytest.mark.parametrize(
-    ["settings_yaml", "env_vars"],
-    [
-        (
-            None,
-            {
-                "PUDL_OUTPUT": "/test/whatever/from/env/output",
-                "PUDL_INPUT": "/test/whatever/from/env/input",
-            },
-        ),
-        (
-            StringIO(
-                yaml.dump(
-                    {
-                        "pudl_in": "/test/workspace",
-                        "pudl_out": "/test/workspace",
-                    }
-                )
-            ),
-            {
-                "PUDL_OUTPUT": "/test/whatever/from/env/output",
-                "PUDL_INPUT": "/test/whatever/from/env/input",
-            },
-        ),
-        (
-            StringIO(
-                yaml.dump(
-                    {
-                        "pudl_in": "/test/workspace",
-                        "pudl_out": "/test/workspace",
-                    }
-                )
-            ),
-            {
-                "PUDL_OUTPUT": "/test/whatever/from/env/different_output",
-                "PUDL_INPUT": "/test/whatever/from/env/different_input",
-            },
-        ),
-    ],
-)
-def test_get_defaults_in_test_environment_use_env_vars(settings_yaml, env_vars):
-    workspace = pathlib.Path(env_vars["PUDL_OUTPUT"]).parent
-    os.environ |= env_vars
-
-    settings = get_defaults(yaml_file=settings_yaml)
-
-    expected_values = {
-        "pudl_in": f"{workspace}",
-        "pudl_out": env_vars["PUDL_OUTPUT"],
-        "data_dir": env_vars["PUDL_INPUT"],
-    }
-
-    for key, value in expected_values.items():
-        assert (key, settings[key]) == (key, value)
-
-    assert os.getenv("PUDL_OUTPUT") == env_vars["PUDL_OUTPUT"]
-    assert os.getenv("PUDL_INPUT") == env_vars["PUDL_INPUT"]
-
-
-@pytest.mark.parametrize(
-    "settings_yaml",
-    [
-        None,
-        StringIO(
-            yaml.dump(
-                {
-                    "pudl_in": "/test/workspace",
-                    "pudl_out": "/test/workspace",
-                }
-            )
-        ),
-    ],
-)
-def test_get_defaults_in_test_environment_use_env_vars_tmpdir(
```

Comment:
I wonder if we could consider a decoupling such that we have one method `set_path_overrides(**kwargs)` to potentially set overrides (which would assign to env variables), and  `get_defaults()` would not have any override functionality at all.

It seems that situations where we need to actually override these are very infrequent and perhaps we could completely get rid of this, assuming that we can set env variables ahead of time. I think the current complexity is probably residue of the merging of these various sources of truth and now there's only one left anyways.

Reply:
IIRC in #2293 we were envisioning using the [Pydantic BaseSettings](https://docs.pydantic.dev/latest/usage/settings/) model instead of the ubiquitous `pudl_settings` dictionary that's passed around everywhere now. Pydantic settings provide a well defined system for overriding defaults including using envvars.

I think the main cases in which we were overriding these paths is in the tests, where we want to separate the test impacts from the user's working environment.

Reply:
Hmm, yeah, Pydantic `BaseSettings` look promising. I will see if I could switch to that as part of this.

Reply:
Conversion to pydantic settings is now done.